### PR TITLE
fs: filter flags ability to read from stdin - fixes #4034

### LINKED ISSUE #4034: Make filter flags (eg `--files-from`) understand `-` as a file name to mean pipe from stdin
As discussed in the below thread, it would be convenient from a scripting perspective to allow any flag that contains a list of files to process to pass that list in via standard in with a dash as a parameter instead of creating intermediate files.  These specifically right now.

      --exclude-from stringArray             Read exclude patterns from file
      --files-from stringArray               Read list of source-file names from file

Using this as an example:
```
find ./ -name 'search_term*' -type f ! -newer /tmp/mytime | sed 's/^.\///' |
rclone delete --dry-run -files-from - remote:
```

Reference topics:
https://forum.rclone.org/t/flag-for-disabling-concurrent-checks-copy/14676/7
https://forum.rclone.org/t/pass-find-results-to-rclone-delete-command/14538/3

### PR DESCRIPTION
**Pull request by fishbullet**
#### What is the purpose of this change?

Add filter flags ability to read from stdin

#### Was the change discussed in an issue or in the forum before?

Related forum discussion 
https://forum.rclone.org/t/pass-find-results-to-rclone-delete-command/14538
Related issue https://github.com/rclone/rclone/issues/4034

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

/cc @ncw

### CODE REVIEW

File: docs/content/flags.md [hunk 1]
```diff
@@ -41,13 +41,13 @@ These flags are available for every command.
       --dump-bodies                          Dump HTTP headers and bodies - may contain sensitive info
       --dump-headers                         Dump HTTP headers - may contain sensitive info
       --exclude stringArray                  Exclude files matching pattern
-      --exclude-from stringArray             Read exclude patterns from file
+      --exclude-from stringArray             Read exclude patterns from file (use - to read from stdin).
```
Comment (ncw) [inline docs/content/flags.md]: This file is auto generated. You'll need to change the original here

https://github.com/rclone/rclone/blob/37d5e75a569e063aa181f440e08ccc9e0cb1567f/fs/filter/filterflags/filterflags.go#L25

File: docs/content/filtering.md [hunk 2]
```diff
@@ -17,7 +17,7 @@ Each path as it passes through rclone is matched against the include
 and exclude rules like `--include`, `--exclude`, `--include-from`,
 `--exclude-from`, `--filter`, or `--filter-from`. The simplest way to
 try them out is using the `ls` command, or `--dry-run` together with
-`-v`.
+`-v`. Filtering understand `-` as a file name to mean pipe from stdin.
```
Comment (ncw) [inline docs/content/filtering.md]: Can you expand this a bit more, eg `--filter-from`, `--exclude-from`, `--include-from`, `--files-from` understand `-` as a file name to mean read from standard input.

**Comment by fishbullet** [comment]
> Looks good - a nice small change.
> 
> Some small comments inline.
> 
> Can you think of a test?

Hey @ncw !

Thank for quick response !

Done documentation changes. 

> Can you think of a test?

I was thinking about test but there are good tests for  `TestFilterForEachLine`
https://github.com/rclone/rclone/blob/master/fs/filter/filter_test.go#L520

And I unable to find more flag tests. Have you any thoughts?

**Comment by ncw** [comment]
> Done documentation changes.

Great!

> > Can you think of a test?
> 
> I was thinking about test but there are good tests for `TestFilterForEachLine`
> https://github.com/rclone/rclone/blob/master/fs/filter/filter_test.go#L520
> 
> And I unable to find more flag tests. Have you any thoughts where I can add some tests for tis feature?

I'd do something like this (untested!) which re-uses the same test, but temporarily patches `os.Stdin` for the duration.

```diff
--- a/fs/filter/filter_test.go
+++ b/fs/filter/filter_test.go
@@ -517,7 +517,7 @@ func TestFilterAddDirRuleOrFileRule(t *testing.T) {
 	}
 }
 
-func TestFilterForEachLine(t *testing.T) {
+func testFilterForEachLine(t *testing.T, useStdin bool) {
 	file := testFile(t, `; comment
 one
 # another comment
@@ -534,7 +534,20 @@ five
 		require.NoError(t, err)
 	}()
 	lines := []string{}
-	err := forEachLine(file, func(s string) error {
+	fileName := file
+	if useStdin {
+		in, err := os.Open(file)
+		require.NoError(t, err)
+		oldStdin := os.Stdin
+		os.Stdin = in
+		defer func() {
+			os.Stdin = oldStdin
+			_ = in.Close()
+		}()
+		fileName = "-"
+
+	}
+	err := forEachLine(fileName, func(s string) error {
 		lines = append(lines, s)
 		return nil
 	})
@@ -542,6 +555,14 @@ five
 	assert.Equal(t, "one,two,three,four,five,six", strings.Join(lines, ","))
 }
 
+func TestFilterForEachLine(t *testing.T) {
+	testFilterForEachLine(t, false)
+}
+
+func TestFilterForEachLineStdin(t *testing.T) {
+	testFilterForEachLine(t, true)
+}
+
 func TestFilterMatchesFromDocs(t *testing.T) {
 	for _, test := range []struct {
 		glob       string
```

**Comment by fishbullet** [comment]
> > Done documentation changes.
> 
> Great!
> 
> > > Can you think of a test?
> > 
> > 
> > I was thinking about test but there are good tests for `TestFilterForEachLine`
> > https://github.com/rclone/rclone/blob/master/fs/filter/filter_test.go#L520
> > And I unable to find more flag tests. Have you any thoughts where I can add some tests for tis feature?
> 
> I'd do something like this (untested!) which re-uses the same test, but temporarily patches `os.Stdin` for the duration.
> 
> ```diff
> --- a/fs/filter/filter_test.go
> +++ b/fs/filter/filter_test.go
> @@ -517,7 +517,7 @@ func TestFilterAddDirRuleOrFileRule(t *testing.T) {
>  	}
>  }
>  
> -func TestFilterForEachLine(t *testing.T) {
> +func testFilterForEachLine(t *testing.T, useStdin bool) {
>  	file := testFile(t, `; comment
>  one
>  # another comment
> @@ -534,7 +534,20 @@ five
>  		require.NoError(t, err)
>  	}()
>  	lines := []string{}
> -	err := forEachLine(file, func(s string) error {
> +	fileName := file
> +	if useStdin {
> +		in, err := os.Open(file)
> +		require.NoError(t, err)
> +		oldStdin := os.Stdin
> +		os.Stdin = in
> +		defer func() {
> +			os.Stdin = oldStdin
> +			_ = in.Close()
> +		}()
> +		fileName = "-"
> +
> +	}
> +	err := forEachLine(fileName, func(s string) error {
>  		lines = append(lines, s)
>  		return nil
>  	})
> @@ -542,6 +555,14 @@ five
>  	assert.Equal(t, "one,two,three,four,five,six", strings.Join(lines, ","))
>  }
>  
> +func TestFilterForEachLine(t *testing.T) {
> +	testFilterForEachLine(t, false)
> +}
> +
> +func TestFilterForEachLineStdin(t *testing.T) {
> +	testFilterForEachLine(t, true)
> +}
> +
>  func TestFilterMatchesFromDocs(t *testing.T) {
>  	for _, test := range []struct {
>  		glob       string
> ```

Neat! 

Updated. 

:+1:

**Comment by ncw** [comment]
That looks great - thank you! I'll merge it now.

**Pull request merged**